### PR TITLE
Extract resources apiserver parts into the apiserver package

### PR DESCRIPTION
**Pull request by mjs**
## Description of change

This PR is part of the cleanup of the apiserver facade registration. All the resources related apiserver code has been moved to the apiserver package, include HTTP endpoints. This means that resources related APIs are now defined in the same way as all other APIs.

Some highlights:

* The Resources and ResourceHookContext facades now exist under apiserver.
* Some mime handling backport code for Go < 1.6 has been removed. We require at least Go 1.7 now.
* Client and unit HTTP endpoints for resource upload and download have been moved to the apiserver package.
* Many layers of completely unnecessary and confusing abstraction have been removed.
* Removed the now-unused apiserver plumbing to support for the component architecture. Both the payloads and resources features have been extracted to apiserver now.
* Drive-by test fixes for Go 1.8.

This PR results in a net removal of over 1700 lines of code, for no functional change!

## QA steps

```
$ juju deploy cs:~cmars/mattermost
# Confirmed that deploy works and resource downloads on to unit.

$ juju list-resources mattermost                          
[Service]
Resource  Supplied by  Revision
bdist     charmstore   3

# Now test with the dummy-resource charm from the QA team

$ cd ~/canonical/repository/charms/dummy-resource
$ juju deploy . --resource foo=foo.txt --resource bar=bar.txt
# Charm emits resource info as status. Confirmed that resource was deployed.

$ juju status | grep Path:
dummy-resource/0*  maintenance  idle   1        10.0.8.199             Path: /var/lib/juju/agents/unit-dummy-resource-0/resources/bar/bar.txt Time: 0 sec Size: 17 bytes

# Update the resource
$ juju attach dummy-resource bar=foo.txt

$ Confirm that the resource is updated
$ juju status | grep Path:
dummy-resource/0*  maintenance  idle   1        10.0.8.199             Path: /var/lib/juju/agents/unit-dummy-resource-0/resources/bar/bar.txt Time: 0 sec Size: 27 bytes
# (note the change in size)

$ juju list-resources dummy-resource
[Service]
Resource  Supplied by  Revision
foo       admin        2017-20-03T00:03
bar       admin        2017-20-03T00:13
```

## Documentation changes

No user visible effects.

## Bug reference

N.A.

### CODE REVIEW

File: apiserver/resources/facade.go [hunk 1]
```diff
@@ -50,13 +56,34 @@ type CharmStore interface {
...
 	// store is the data source for the facade.
-	store resourceInfoStore
+	store backend
 
 	newCharmstoreClient func() (CharmStore, error)
 }
 
-// NewFacade returns a new resoures facade for the given Juju state.
-func NewFacade(store DataStore, newClient func() (CharmStore, error)) (*Facade, error) {
+// NewPublicFacade creates apublic API facade for resources. It is
```
Comment (wallyworld) [inline apiserver/resources/facade.go]: typo

File: apiserver/resources/facade.go [hunk 2]
```diff
@@ -50,13 +56,34 @@ type CharmStore interface {
...
+	}
+	facade, err := NewFacade(rst, newClient)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return facade, nil
+}
+
+// NewFacade returns a new resoures API facade.
+func NewFacade(store backend, newClient func() (CharmStore, error)) (*Facade, error) {
```
Comment (wallyworld) [inline apiserver/resources/facade.go]: Does this really need to be exported? I guess it is used in tests.

File: apiserver/resources/server_addpending_test.go [hunk 3]
```diff
@@ -293,24 +292,24 @@ func (s *AddPendingResourcesSuite) TestWithURLUpload(c *gc.C) {
 //func (s *AddPendingResourcesSuite) TestUnknownResource(c *gc.C) {
```
Comment (wallyworld) [inline apiserver/resources/server_addpending_test.go]: Not sure why this was originally commented out. 
We should either skip with a bug and a todo, or delete.

File: apiserver/resources_mig.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
+// Copyright 2016 Canonical Ltd.
```
Comment (wallyworld) [inline apiserver/resources_mig.go]: 2017 if this is new

File: resource/resourceadapters/opener.go [hunk 5]
```diff
@@ -13,23 +13,40 @@ import (
 	corestate "github.com/juju/juju/state"
 )
 
+// NewResourceOpener returns a new resource.Opener for the given unit.
+//
+// TODO(mjs): This is the entry point for a whole lot of untested shim
+// code in this package. At some point this should be sorted out.
+func NewResourceOpener(st *corestate.State, unitName string) (opener resource.Opener, err error) {
```
Comment (wallyworld) [inline resource/resourceadapters/opener.go]: Can you add a comment that the caller is responsible for closing state just be make it explicit.
I traced to code to be sure we weren't leaking connections, so would be nice to inform the reader the expectation.

File: apiserver/params/resources_unit.go [hunk 6]
```diff
@@ -1,38 +1,31 @@
...
-
-// TODO(ericsnow) Eliminate the apiserver dependencies, if possible.
-
-import (
-	"github.com/juju/juju/apiserver/params"
-	"github.com/juju/juju/resource/api"
-)
+package params
 
 // ListResourcesArgs holds the arguments for an API request to list
```
Comment (mjs) [inline apiserver/params/resources_unit.go]: Fixed

File: apiserver/resources/facade.go (hunk 1 above)
Comment (mjs) [inline apiserver/resources/facade.go]: Fixed

File: apiserver/apiserver.go [hunk 7]
```diff
@@ -475,7 +509,7 @@ func (srv *Server) endpoints() []apihttp.Endpoint {
 		},
 	)
 	add("/migrate/resources",
-		&resourceUploadHandler{
+		&resourceMigUploadHandler{
```
Comment (mjs) [inline apiserver/apiserver.go]: Fixed :)

File: apiserver/resources/facade.go (hunk 2 above)
Comment (mjs) [inline apiserver/resources/facade.go]: Yep, for tests.

File: apiserver/resources_mig.go (hunk 4 above)
Comment (mjs) [inline apiserver/resources_mig.go]: Fixed

File: apiserver/resources/server_addpending_test.go (hunk 3 above)
Comment (mjs) [inline apiserver/resources/server_addpending_test.go]: Not sure what the background here is but it looks like one of the other (enabled) tests already covers this.

File: resource/resourceadapters/opener.go (hunk 5 above)
Comment (mjs) [inline resource/resourceadapters/opener.go]: Done

**Comment by mjs** [comment]
$$merge$$

**Comment by mjs** [comment]
$$merge$$

**Comment by mjs** [comment]
$$merge$$

**Comment by mjs** [comment]
$$mongodb-died$$

**Pull request merged**